File: src/builder.py
from markdown import markdown

from src.posts import process_posts
from src.replacement_tags import CONTENT, ROOT, TITLE, DESCRIPTION, CARD_IMAGE
from src.markdown_processor import markdown_to_html

def get_layout():
    layout_html = open(f'{ROOT}/layout.html','r')
    return layout_html.read()

def get_li_for_post(post):
    return f'<li><a href="post/{post.name}">{post.title}</a><span>{post.date}</span><p>{post.subtitle}</p></li>'

def process_index(posts):
    """
    Creates the index page with the list of posts
    """
    index_html = open(f'{ROOT}/index.html','r').read()

    # Sort by date order
    posts = sorted(posts, key=lambda k: int(k.date.replace('-','')), reverse = True)

    # put content into html template
    content = ''
    for post in posts:
        c = get_li_for_post(post)
        content = f'{content}{c}'
    index_html = index_html.replace(CONTENT, content)

    layout_html = get_layout()
    html_out = layout_html.replace(CONTENT, index_html)

    html_out = html_out.replace(TITLE, 'Divan Visagie')
    html_out = html_out.replace(DESCRIPTION, 'Divan\'s personal blog')
    html_out = html_out.replace(CARD_IMAGE, 'favicon.ico')

    f_out = open(f'{ROOT}/public/index.html', 'w')
    f_out.write(html_out)

def process_simple(content_folder, template_name):
    """
    Process a markdown folder into a site page
    """
    template_html = open(f'{ROOT}/{template_name}.html', 'r').read()
    content_md = open(f'{ROOT}/content/{content_folder}/index.md', 'r').read()
    layout_html = get_layout()
    
    # Put content html inside the template for this page
    content_html = markdown_to_html(content_md)
    content_html = template_html.replace(CONTENT, content_html)

    # Put the content into the site layout template
    html_out = layout_html.replace(CONTENT, content_html)

    html_out = html_out.replace(TITLE, 'Divan Visagie')
    html_out = html_out.replace(DESCRIPTION, 'Divan\'s personal blog')
    html_out = html_out.replace(CARD_IMAGE, 'favicon.ico')
    
    f_out = open(f'{ROOT}/public/{content_folder}/index.html', 'w')
    f_out.write(html_out)

def build():
    print('starting the process')
    process_simple('about', 'about')
    posts = process_posts()
    process_index(posts)

if __name__ == '__main__':
    println('Please run from root')

File: src/builder.js
const path = require('path')
const fs = require('fs-extra')
const urljoin = require('url-join')

const { markdownToHtml } = require('./markdown_processor')

const { CONTENT, TITLE, DESCRIPTION, CARD_IMAGE, OUT_DIR, TEMPLATE_ROOT } = require('./replacement_tags')
const { getPosts, getPostLineItem } = require('./posts')

const rootUrl = 'https:\/\/dvisagie.com'

async function getLayout() {
    return await fs.readFile(
        path.join(TEMPLATE_ROOT, 'layout.html'), 'utf8'
    )
}

function findFirstImage(html) {
    try {
        return html.split('<img')[1]
            .split('src="')[1]
            .split('"')[0]
            .trim()
    } catch {
        return //never go full python
    }
}

/**
 * Process a markdown folder into a site page
 */
async function processFolderWithTemplate(contentFolder, templateName) {
    const metaPath = path.join(`./content`, contentFolder, `index.json`)
    let meta = {};
    if (fs.existsSync(metaPath))
        meta = JSON.parse(await fs.readFile(metaPath, 'utf8'))

    const templatePath = path
        .join(TEMPLATE_ROOT, `${templateName}.html`)
    const markdownPath = path
        .join('./content', contentFolder, 'index.md')

    const templateHtml = await fs.readFile(templatePath, 'utf8')
    const contentMd = await fs.readFile(markdownPath, 'utf8')
    const layoutHtml = await getLayout()

    // Put the content into the site layout template
    let title = ''
    if (meta.title) {
        title = `${meta.title} - `
    }

    const convertedMarkdown = markdownToHtml(contentMd)

    let cardImage = `${rootUrl} / favicon.ico`
    if (!meta.header) {
        const found = findFirstImage(convertedMarkdown)
        if (found)
            cardImage = urljoin(rootUrl, contentFolder, found)
    } else {
        cardImage = urljoin(rootUrl, contentFolder, meta.header)
    }

    let pageContent = templateHtml.replace(CONTENT, convertedMarkdown)
    if (meta.header) {
        pageContent = `<img class="post-header" alt = "An image displayed as a header before the article for decorative purposes." src="${meta.header}" ></img >\n${pageContent}`
    }

    const htmlOut = layoutHtml
        .replace(TITLE, title)
        .split(TITLE).join(meta.title || 'Divan Visagie')
        .split(DESCRIPTION).join(meta.subtitle || `Divan's personal blog`)
        .split(CARD_IMAGE).join(cardImage)
        .replace(CONTENT, pageContent)

    const contentFolderPath =
        path.join('.', OUT_DIR, contentFolder)

    if (!fs.existsSync(contentFolderPath)) {
        await fs.mkdir(contentFolderPath, { recursive: true })
    }

    await fs.writeFile(
        path.join(contentFolderPath, 'index.html'),
        htmlOut
    )

    await fs.copy(
        path.join('./content', contentFolder),
        contentFolderPath
    )

    return meta
}

async function processIndexPage(posts) {
    const layoutHtml = await getLayout()
    const indexHtml = await fs.readFile(
        path.join(TEMPLATE_ROOT, 'index.html')
    )

    posts.sort((a, b) => {
        const da = new Date(a.date)
        const db = new Date(b.date)

        // console.log(da, db)
        return db - da
    })


    const list = `${posts.map(getPostLineItem).join('')}`

    const output = layoutHtml
        .split(TITLE).join('')
        .split(DESCRIPTION).join(`Divan's personal blog`)
        .replace(CONTENT, indexHtml)
        .replace(CONTENT, list)

    await fs.writeFile('./build/index.html', output)
}

async function buildPages() {
    const timer = 'Pages built'
    console.log('Starting page build process')
    console.time(timer)
    await processFolderWithTemplate('about', 'about')
    await processFolderWithTemplate('cv', 'about')

    const posts = []
    for (let post of await getPosts()) {
        const postMetadata = await processFolderWithTemplate(`post/${post}`, 'post')
        postMetadata.name = post
        posts.push(postMetadata)
    }

    await processIndexPage(posts)

    console.timeEnd(timer)
}


module.exports = { buildPages }